src/vcpkg/sourceparagraph.cpp

```diff
@@ -1449,18 +1454,52 @@ namespace vcpkg
         {
             if (!error_info->expected_types.empty())
             {
-                print2(Color::error,
-                       "Error: There are invalid field types in the CONTROL or manifest file of ",
-                       error_info->name,
-                       '\n');
-                print2("The following fields had the wrong types:\n\n");
+                Strings::append(message,
+                                "Error: There are invalid field types in the CONTROL or manifest file of ",
+                                error_info->name,
+                                '\n');
+                Strings::append(message, "The following fields had the wrong types:\n\n");
 
                 for (const auto& pr : error_info->expected_types)
                 {
-                    vcpkg::printf("    %s was expected to be %s\n", pr.first, pr.second);
+                    Strings::append(message, "    ", pr.first, " was expected to be ", pr.second, "\n");
                 }
-                print2("\n");
+                Strings::append(message, "\n");
+            }
+        }
+
+        return message;
+    }
+
+    void print_error_message(Span<const std::unique_ptr<ParseControlErrorInfo>> error_info_list)
+    {
+        auto msg = ParseControlErrorInfo::format_errors(error_info_list);
+
+        // To preserve previous behavior, each line starting with "Error" should be error-colored. All other lines
+        // should be neutral color.
+
+        // To minimize the number of print calls on Windows (which is a significant performance bottleneck), this
+        // algorithm chunks groups of similarly-colored lines.
+        auto lines = Strings::split(msg, '\n');
```

Comment:
I'm ... really not a big fan of this loop:

* allocating N strings and at least 2N-3N space in the name of perf which is not an obvious win to me
* not a fan of "juggling" back and forth between `i` and `j` because it makes seeing that this loop is linear difficult
* this technically alters the output because repeated newlines are collapsed which is probably not intended

How about one of these?

A version where the first line search block is "rotated around" to ensure a run of newlines at the beginning are printed.

```
const char* first = msg.data();
const char* const last = first + msg.size();
static constexpr StringLiteral error = "Error";
while (first != last) { // for each alternating error/not-error block
    // find the block of lines while each starts with Error:
    auto block_start = first;
    for (;;) {
        first = std::find_if_not(first, last, is_newline);
        if (!Strings::starts_with(first, last, error)) {
            break;
        }

        first = std::find_if(first, last, is_newline);
    }

    if (block_start != first) {
        print2(Colors::error, StringView(block_start, first));
    }

    if (first == last) {
        break;
    }

    // note: first points to the start of a line not beginning with Error: here
    // find the block of lines while each does NOT start with Error:
    block_start = first;
    do {
        first = std::find_if(first, last, is_newline);
        first = std::find_if_not(first, last, is_newline);
    } while (!Strings::starts_with(first, last, error));

    if (block_start != first) {
        print2(Colors::message, StringView(block_start, first));
    }
}
```

A version which establishes simpler loop invariants to let that inner loop be easier to read:

```
const char* first = msg.data();
const char* const last = first + msg.size();
static constexpr StringLiteral error = "Error";
if (!Strings::starts_with(first, last, error)) {
    auto block_start = first;
    auto first = Strings::find(first, last, "\nError"); // I'm not sure if this find exists but you get the point
    if (first == last) {
        print2(msg);
        return;
    }

    ++first; // skip the \n
    print2(StringView(block_start, first));
}

do { // for each alternating error/not-error block
    assert(Strings::starts_with(first, last, error));
    auto block_start = first;
    do {
        first = std::find_if(first, last, is_newline);
        first = std::find_if_not(first, last, is_newline);
    } while (Strings::starts_with(first, last, error));
    print2(Colors::error, StringView(block_start, first));
    if (first == last) {
        return
    }

    // note: !Strings::starts_with(first, last, error) && first != last
    block_start = first;
    do {
        first = std::find_if(first, last, is_newline);
        first = std::find_if_not(first, last, is_newline);
    } while (!Strings::starts_with(first, last, error));
    print2(StringView(block_start, first));
} while (first != last);
```

Reply:
I've rewritten the algorithm to perform zero allocations, does this iteration work for you?

Reply:
LGTM